Draft: natively support fmks coordinates

gnwong opened the pull request:
The commits in this pull request implement support for FMKS coordinates. Since there is not a clear way to go from Kerr-Schild coordinates to FMKS coordinates without performing root finding, the code implements a so-called "sks-map" feature, which provides a general framework for identifying "which [native/simulation] coordinates correspond to a given Spherical Kerr Schild point". The ```sks-map``` variable is a set of 2 two-dimensional arrays that provide the native ```x1``` and ```x2``` coordinates for KS coordinate points regularly spaced in ```r``` and ```theta```. 

There's a lot in this PR, and I would not be surprised if some of the design choices are objectionable.

Right now, the code also isn't fully tested. I will update the text of this PR as I go through the tests and other validation.

### Checklist
- [x] Check that ```sks-map``` implementation gives expected values.
- [x] Check that code-to-cgs conversions give expected values.
- [x] Check that SKS vector components give expected values.
- [ ] Add info to function docstrings for when to/not to treat ```Coordinates::fmks``` as ```Coordiantes::sks```
- [ ] Troubleshoot spurious hot pixels.
- [ ] Troubleshoot large values in "final step".

### Other notes
- ```blacklight``` saves the primitives (including for fluid velocity and magnetic field), interpolates them in regular sks, and then constructs scalars. This procedure will yield different values for things like ```bsq``` (and therefore ```sigma```) compared to ```ipole```, which saves the scalars in fmks and interpolates there. The differences are largest where fmks and sks differ substantially (e.g., near poles) and where the four-vectors vary rapidly. It doesn't seem to be that much of a problem (see images below), but it's good to keep in mind.
- ```blacklight``` uses a different formula for computing ```Theta_e``` given fluid density and total internal energy. The local difference turns out to be a factor of ```2.0 * (2.0 + R) / 3.0 / (1.0 + R)``` for a given ```R = T_i/T_e``` assuming ```game = 4.0/3.0``` and ```gamp = 5.0/3.0```. To produce the comparison, I modified ```ipole```'s method for computing ```Theta_e``` to match ```blacklight```'s, but that change will not be persisted to the public version of ```ipole```.

gnwong committed:
this version reads fmks files and makes images that look reasonable, but it should be checked.

gnwong committed:
update with test output hooks

gnwong commented:
Saved for reference — two images produced with ipole and blacklight running on the same input GRMHD file using the rlow/rhigh electron temperature model synced to blacklight's implementation (see caveats above).

## ipole image

```./ipole --freqcgs=230.e9 --MBH=6.2e9 --thetacam=170 --dump=Sa+0.94_288/dump_00000900.h5 --outfile=ipole_image.h5 --nx=160 --ny=160 --fov=159.37698 --M_unit=7.7754e+28```
![image](https://github.com/c-white/blacklight/assets/3887665/04f385cd-c33a-4336-ad6a-29be17c8a5c8)

## blacklight image

```# relevant blacklight parameters (others default ~ consistent with example)
# Simulation parameters
simulation_format       = iharm3d           # format of GRMHD data
simulation_file         = Sa+0.94_288/dump_00000900.h5
simulation_coord        = fmks              # simulation coordinates (sks, cks)
simulation_a            = 0.9375            # dimensionless black hole spin
simulation_m_msun       = 6.2e9            # black hole mass in solar masses
simulation_rho_cgs      = 1.0e-16          # unit code density in g/cm^3

# Plasma parameters
plasma_mu         = 0.5         # molecular weight of fluid in proton masses
plasma_ne_ni      = 1.0         # electrons per ion
plasma_model      = ti_te_beta  # thermal electron temperature model (ti_te_beta, code_kappa)
plasma_rat_low    = 1.0         # ion-electron temperature ratio at zero plasma beta
plasma_rat_high   = 40.0        # ion-electron temperature ratio at infinite plasma beta

# Camera parameters
camera_th         = 170.0   # KS polar coordinate theta in degrees (0/180 for N/S)
camera_width      = 44.0   # full width of image in gravitational units
camera_resolution = 160     # number of pixels per side
```
![blacklight_image](https://github.com/c-white/blacklight/assets/3887665/ee019678-e78d-4351-88a7-53120907d744)

c-white committed:
changed ordering of coordinate systems throughout code

c-white committed:
removed trailing whitespace

c-white committed:
removed temporary comments and fixed too-long lines

c-white committed:
reordered function definitions

c-white committed:
changed from references to pointers for input POD to be overwritten

c-white committed:
added return statements to void functions

c-white committed:
changed logical operators to word forms

c-white committed:
made bracing style consistent

c-white committed:
made various style changes and fixed bug in indexing FMKS data without interpolation

c-white committed:
fixed miscellaneous typos and avoided warnings

c-white commented:
This looks great! I touched up the style, hopefully without breaking anything. As far as I'm concerned, the only things needed before merging are:

- [x] Understanding and removing the remaining TODO comments.
- [x] Rerunning an FMKS job just to make sure it still looks the same after my changes.

Optional tasks before merge:

- [ ] Updating the wiki with FMKS.
- [ ] Understanding hot pixels.
- [ ] Understanding the final step issue above.
- [x] Adding a flag for Blacklight to match ipole's Theta_e behavior.

gnwong committed:
process easy-to-check TODO statements and add independent ion/electon adiabatic index machinery for iharm3d files

gnwong committed:
add use_ipole_tpte flag to reproduce ipole energy partition behavior.

gnwong commented:
Responding to items in @c-white's most recent comment since I cannot edit it:

- [x] I've rerun the original FMKS job and things look identical.
- [ ] In processing the TODO's I have left the one you added in ```radiation_integrator/simulation_sampling.cpp```. To be honest I am not 100% sure what I was going for there originally, but it had to do with using ```sks_map``` rather than only using ```x[123]v``` values. The note that it could be more permissive was to do with how the (current) infrastructure (does not) deal with whenever the requested coordinate might lie outside of ```sks_map```.
- [ ] I've added a flag for Blacklight to match ipole's Theta_e behavior. The new ```use_ipole_tpte``` flag (along with associated adiabatic_gamma_* machinery) reproduces the values of ```theta_e``` that ipole does. Comparing image morphologies shows strong consistency, although the "smoothing" differences seen between blacklight and ipole previously persist and seem to change total intensity in some pixels (see comparison images posted in next comment). Should figure out why this is so inconsistent and decide if it is a problem.

Other points:

 - [ ]  Understanding hot pixels.
 - [ ]  Understanding the final step issue above.
 - [ ] Should the ```sks_map_n[12]``` be something the user could specify? I think the 2048 default values are probably perfectly reasonable, but they could in principle be overwritten at runtime on user choice.
 - [ ]  Updating the wiki with FMKS. I would be happy to draft something but I am not sure what kind of workflow you'd like.

gnwong commented:
Images produced using the original ipole implementation of ```theta_e``` from summing the fluid internal energies.

Blacklight image
![blacklight_image_ipole_tpte](https://github.com/c-white/blacklight/assets/3887665/722ef345-ecf1-485f-891c-aa7115413b12)

ipole image
![image_ipole_tpte](https://github.com/c-white/blacklight/assets/3887665/6dc733d7-4f8b-46d5-a1d5-e84fd1ad2138)

c-white committed:
Synced code relating to two-fluid thermodynamics:

Now any data file can be run with either plasma_use_p=true (original
default), in which T_e is calculated assuming p_i + p_e = p from the
simulation, or plasma_use_p=false, in which T_e is calculated assuming
u_i + u_e = u from the simulation. The latter case requires knowing the
adiabatic indices for the species separately.

ipole uses the latter formula, though also assumes in the T_e
calculation that n_e/n_i = 1 and mu = 1/2. Here, these parameters always
have the values given by the user in the input file.

The code attempts to be sensible regarding input adiabatic indices. When
they are not needed at all, they are generally ignored, with warnings
issued if they are provided. In the case of Athena++ files, values for
plasma_gamma, when not needed, are silently ignored. Required values
will result in an error if not provided. Sometimes the values are
optional -- the file format must have them, and a warning is issued if
the user provides values and those values do not match the file values.
In some cases the file format might provide them, in which case there is
a warning if inconsistency is found, and there is an error of no source
can be found.

c-white committed:
removed remaining TODO notes

c-white committed:
fixed bug in reading adiabatic indices
